src/mscorlib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -129,6 +129,21 @@ private struct Entry {
             }
         }
 
+        public Dictionary(IEnumerable<KeyValuePair<TKey, TValue>> collection):
+            this(collection, null) { }
+
+        public Dictionary(IEnumerable<KeyValuePair<TKey, TValue>> collection, IEqualityComparer<TKey> comparer):
+            this(comparer)
+        {
+            if (collection == null) {
```

Comment:
Do we want to try casting to ICollection, and if so, get .Count in order to pre-size the dictionary?